blockchain/state_manager/src/lib.rs

```diff
@@ -54,6 +56,99 @@ const DEFAULT_TIPSET_CACHE_SIZE: NonZeroUsize =
 /// Intermediary for retrieving state objects and updating actor states.
 type CidPair = (Cid, Cid);
 
+// Various structures for implementing the tipset state cache
+
+struct TipsetStateCacheInner {
+    values: LruCache<TipsetKeys, CidPair>,
+    pending: Vec<(TipsetKeys, Arc<TokioMutex<()>>)>,
+}
+
+impl Default for TipsetStateCacheInner {
+    fn default() -> Self {
+        Self {
+            values: LruCache::new(DEFAULT_TIPSET_CACHE_SIZE),
+            pending: Vec::with_capacity(8),
+        }
+    }
+}
+
+struct TipsetStateCache {
+    cache: Arc<StdMutex<TipsetStateCacheInner>>,
+}
+
+enum Status {
+    Done(CidPair),
+    Empty(Arc<TokioMutex<()>>),
+}
+
+impl TipsetStateCache {
+    pub fn new() -> Self {
+        Self {
+            cache: Arc::new(StdMutex::new(TipsetStateCacheInner::default())),
+        }
+    }
+
+    fn with_inner<F, T>(&self, func: F) -> T
+    where
+        F: FnOnce(&mut TipsetStateCacheInner) -> T,
+    {
+        let mut lock = self.cache.lock().unwrap();
+        func(&mut lock)
+    }
+
+    pub async fn get_or_else<F, Fut>(&self, key: &TipsetKeys, cb: F) -> anyhow::Result<CidPair>
+    where
+        F: Fn(TipsetKeys) -> Fut,
```

Comment:
We can skip the key argument here, can't we?

```suggestion
        F: Fn() -> Fut,
```